zebra-network/src/peer/connection.rs

```diff
@@ -368,455 +373,390 @@ where
-                                _ => unreachable!("unexpected state after AwaitingResponse: {:?}, peer_msg: {:?}, client_receiver: {:?}",
-                                                  self.state,
-                                                  peer_msg,
-                                                  self.client_rx,
-                                ),
-                            };
-                            // If the message was not consumed, check whether it
-                            // should be handled as a request.
-                            if let Some(msg) = request_msg {
-                                // do NOT instrument with the request span, this is
-                                // independent work
-                                self.handle_message_as_request(msg).await;
-                            } else {
-                                // Otherwise, check whether the handler is finished
-                                // processing messages and update the state.
-                                self.state = match self.state {
-                                    State::AwaitingResponse {
-                                        handler: Handler::Finished(response),
-                                        tx,
-                                        ..
-                                    } => {
-                                        let _ = tx.send(response.map_err(Into::into));
-                                        State::AwaitingRequest
-                                    }
-                                    pending @ State::AwaitingResponse { .. } => pending,
-                                    _ => unreachable!(
-                                        "unexpected failed connection state while AwaitingResponse: client_receiver: {:?}",
-                                        self.client_rx
-                                    ),
-                                };
-                            }
-                        }
-                        Either::Right((Either::Left(_), _peer_fut)) => {
-                            trace!(parent: &span, "client request timed out");
-                            let e = PeerError::ClientRequestTimeout;
-                            self.state = match self.state {
-                                // Special case: ping timeouts fail the connection.
-                                State::AwaitingResponse {
-                                    handler: Handler::Ping(_),
-                                    ..
-                                } => {
-                                    self.fail_with(e);
-                                    State::Failed
+            }
+            // We're awaiting a response to a client request,
+            // so wait on either a peer message, or on a request cancellation.
+            State::AwaitingResponse {
+                span,
+                mut tx,
+                mut handler,
+                request_timer,
```

Comment:
When does the timer start and finish, compared with the old timer?

Reply:
The timer should be constructed at the end of a state transition, same as before. It used to happen in the body of `handle_client_request`, once we've finished processing the request and manually updated the various bits of state to approximate a state transition. Now it is handled in the `TryFrom` impl where we take the `Transition` and use it to construct the subsequent `State`, as far as I can tell this should result in it starting and finishing at the same time.